3.evaluate_model/scripts/nbconverted/get_LOIO_probabilities.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import pathlib
+import warnings
+
+import pandas as pd
+
+from sklearn.linear_model import LogisticRegression
+from sklearn.utils import parallel_backend
+from sklearn.exceptions import ConvergenceWarning
+from joblib import load
+
+import sys
+
+sys.path.append("../utils")
+from split_utils import get_features_data
+from train_utils import get_dataset, get_X_y_data
+
+
+# In[2]:
+
+
+# load labeled data
+labeled_data_path = pathlib.Path("../0.download_data/data/labeled_data.csv.gz")
+labeled_data = get_features_data(labeled_data_path)
+
+# preview labeled data
+print(labeled_data.shape)
+labeled_data.head(5)
+
+
+# In[3]:
+
+
+# see number of images to
+num_images = labeled_data["Metadata_DNA"].unique().shape[0]
+print(f"There are {num_images} images to perform LOIO evaluation on per model.")
+
+
+# In[4]:
+
+
+# directory to load the models from
+models_dir = pathlib.Path("../2.train_model/models/")
+
+# use a list to keep track of LOIO probabilities in tidy long format for each model combination
+compiled_LOIO_wide_data = []
+
+count = 0
+
+# iterate through each model (final model, shuffled baseline model, etc)
+# sorted so final models are loaded before shuffled_baseline
+for model_path in sorted(models_dir.iterdir()):
+    # only perform LOIO with hyper params from final models
+    if "shuffled" in model_path.name:
+        continue
+
+    # load the model
+    model = load(model_path)
+    # determine feature type from model file name
+    feature_type = model_path.name.split("__")[1].replace(".joblib", "")
+
+    print(
+        f"Performing LOIO for feature type {feature_type} with parameters C: {model.C}, l1_ratio: {model.l1_ratio}"
+    )
+
+    # iterate through image paths
+    for image_path in labeled_data["Metadata_DNA"].unique():
```

Comment:
Seeing the added complexity of a nested for-loop, consider iterating through `image_path` (and others) using [`zip`](https://docs.python.org/3/library/functions.html#zip) or similar approaches without for-loop nesting. Doing this might require rework, but could help with troubleshooting, maintenance, and readability. For example:

```python
list_one = ["one", "two", "three"]
list_two = ["dog","cat","bat"]
list_three = [1, 2, 3]

for one, two, three in zip(list_one, list_two, list_three):
    print(one, two, three)
```

Reply:
I have never heard of this tip, thank you for sharing!
However, I think it might not make sense in this context considering that there are only 2 for loops (so it is decently readable) and the model needs to be loaded in the first for loop but not every iteration of the second for loop (so this strategy might increase runtime).